ml-agents/mlagents/trainers/buffer.py

```diff
@@ -12,285 +13,277 @@ class BufferException(UnityException):
+    ) -> "AgentBuffer":
+        """
+        Creates a mini-batch from a random start and end.
+        :param batch_size: number of elements to withdraw.
+        :param sequence_length: Length of sequences to sample.
+            Number of sequences to sample will be batch_size/sequence_length.
+        """
+        num_seq_to_sample = batch_size // sequence_length
+        mini_batch = AgentBuffer()
+        buff_len = len(next(iter(self.values())))
+        num_sequences_in_buffer = buff_len // sequence_length
+        start_idxes = (
+            np.random.randint(num_sequences_in_buffer, size=num_seq_to_sample)
+            * sequence_length
+        )  # Sample random sequence starts
+        for i in start_idxes:
+            for key in self:
+                mini_batch[key].extend(self[key][i : i + sequence_length])
+        return mini_batch
+
+    def save_to_file(self, file_object: BinaryIO) -> None:
+        """
+        Saves the AgentBuffer to a file-like object.
+        """
+        with h5py.File(file_object) as write_file:
+            for key, data in self.items():
+                write_file.create_dataset(key, data=data, dtype="f", compression="gzip")
 
-    def reset_update_buffer(self):
+    def load_from_file(self, file_object: BinaryIO) -> None:
         """
-        Resets the update buffer
+        Loads the AgentBuffer from a file-like object.
         """
-        self.update_buffer.reset_agent()
+        with h5py.File(file_object) as read_file:
+            for key in list(read_file.keys()):
+                self[key] = AgentBuffer.AgentBufferField()
+                # extend() will convert the numpy array's first dimension into list
+                self[key].extend(read_file[key][()])
 
-    def truncate_update_buffer(self, max_length, sequence_length=1):
+    def truncate(self, max_length: int, sequence_length: int = 1) -> None:
         """
-        Truncates the update buffer to a certain length.
+        Truncates the buffer to a certain length.
 
         This can be slow for large buffers. We compensate by cutting further than we need to, so that
         we're not truncating at each update. Note that we must truncate an integer number of sequence_lengths
         param: max_length: The length at which to truncate the buffer.
         """
-        current_length = len(next(iter(self.update_buffer.values())))
+        current_length = len(next(iter(self.values())))
         # make max_length an integer number of sequence_lengths
         max_length -= max_length % sequence_length
         if current_length > max_length:
-            for _key in self.update_buffer.keys():
-                self.update_buffer[_key] = self.update_buffer[_key][
-                    current_length - max_length :
-                ]
+            for _key in self.keys():
+                self[_key] = self[_key][current_length - max_length :]
+
+
+class AgentProcessorBuffer(dict):
```

Comment:
Can we use separate files? I think it would reduce some confusion.

Reply:
Split into separate files -> `agent_processor.py` contains the processing buffer